### PR TITLE
feat(autoware_adapi_v1_msgs): add object recognition msg

### PR DESCRIPTION
**Pull request by tkhmy**
## Description
- add the msg required for object recognition

## Related links
https://github.com/autowarefoundation/autoware/issues/3225#issuecomment-1409895712
https://github.com/autowarefoundation/autoware-documentation/pull/321

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/perception/msg/Shape.msg:1]: Could you avoid using zero or assign `UNKNOWN` to distinguish from uninitialized value?

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:1]: you mean 0 is used for `UNKNOWN` purpose?
I can change it

Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:1]: > you mean 0 is used for UNKNOWN purpose?

Yes.

> This message is refer from here

Constants do not have to be the same as the internal messages, nor should we expect them to be. In the API adapter, even if the values are the same, they need to be converted using a switch statement, etc., for future changes.

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg]: changed at [bbbfc45](https://github.com/autowarefoundation/autoware_adapi_msgs/pull/25/commits/bbbfc45df9dc751fa5513bdc421791cd022f60fc)

**Comment by isamu-takagi** [comment]
@tkhmy 
LGTM. The remaining consideration is whether to use generic shapes (https://github.com/autowarefoundation/autoware_adapi_msgs/pull/26).

File: autoware_adapi_v1_msgs/perception/msg/PredictedPath.msg
Comment (miursh) [inline autoware_adapi_v1_msgs/perception/msg/PredictedPath.msg:2]: maybe 'builtin_interfaces/msg/Duration time_step' is missing?
https://github.com/tier4/autoware_auto_msgs/blob/tier4/main/autoware_auto_perception_msgs/msg/PredictedPath.idl#L11

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/PredictedPath.msg:3]: Thank you.
Added back the missing `time_step`
[a9781b5](https://github.com/autowarefoundation/autoware_adapi_msgs/pull/25/commits/a9781b5de6e26b3285a4271f2cd08c59165f5f81)

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/perception/msg/Shape.msg:4]:
```suggestion
uint8 PRISM=3
```

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/perception/msg/Shape.msg:2]:
```suggestion
uint8 BOX=1
```

File: autoware_adapi_v1_msgs/perception/msg/PredictedObject.msg
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/perception/msg/PredictedObject.msg:1]:
```suggestion
unique_identifier_msgs/UUID id
```

File: autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg:5]: I don't think `predicted_paths` will be used much. How about removing it for now and adding it if needed?

File: autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg
Comment (isamu-takagi) [inline autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg:3]:
```suggestion
geometry_msgs/PoseWithCovariance pose
geometry_msgs/TwistWithCovariance twist
geometry_msgs/AccelWithCovariance accel
```

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:4]: from autoware_auto_perception_msgs, the shape is using `POLYGON`, I think to prevent confuse, we should keep the same?

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:2]: Same question with `POLYGON`

File: autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg:3]: for this I think we can shorten to `pose_with_covariance`, to differentiate with pose.
How do you think?

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:4]: This actually means PRISM (it uses `dimension.z` as height). I think it's better to change `autoware_auto_perception_msgs` to avoid confusion.

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:2]: I referred to [shape_msgs](https://github.com/ros2/common_interfaces/blob/rolling/shape_msgs/msg/SolidPrimitive.msg). Either one is fine.

File: autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg
Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg:3]: In [geometry_msgs](https://github.com/ros2/common_interfaces/blob/rolling/geometry_msgs/msg/PoseWithCovarianceStamped.msg), PoseStamped and PoseWithCovariance are also simply use `pose`, so I don't think there is much need to distinguish them.

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:4]: I see, do you mean we also need to change the autoware_auto_msgs?

File: autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg:3]: I see, okay, I will change it.

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (isamu-takagi) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:4]: Not needed right now as it affects some nodes.

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg:4]: okay, then I just change in api_msgs

File: autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/PredictedObjectKinematics.msg]: changed at 
f6bfa5e

File: autoware_adapi_v1_msgs/perception/msg/PredictedObject.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/PredictedObject.msg]: changed at 
f6bfa5e

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg]: changed at 
f6bfa5e

File: autoware_adapi_v1_msgs/perception/msg/Shape.msg
Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/Shape.msg]: changed at 
f6bfa5e

**Comment by isamu-takagi** [comment]
@tkhmy `PredictedObject` is the classification that the perception component uses internally, so please rename it to `DynamicObject`.
@yukkysaito I think that `classification` and `shape` are also used in static object, so I didn't add the `dynamic` prefix, is this right?

| Before | After |
| ---- | ---- |
| ObjectClassification | ObjectClassification |
| ObjectKinematics | DynamicObjectKinematics |
| PredictedObject | DynamicObject |
| PredictedObjectArray | DynamicObjectArray |
| PredictedPath | DynamicObjectPath
| Shape | ObjectShape |

**Comment by yukkysaito** [comment]
It depends on the definition of DynamicObject, but I understand that the current Autoware treats anything that was not there when the map was created as a DynamicObject (including cones). So I think it is ok to put Dynamic as prefix. cc @xmfcx @mitsudome-r

**Comment by tkhmy** [comment]
Thanks for the commend
I will change it

File: autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg
Comment (xmfcx) [inline autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: I think these `XWithCovariance` messages are Too large.

For example [Pose](https://github.com/ros2/common_interfaces/blob/rolling/geometry_msgs/msg/Pose.msg) message contains a Point and a Quaternion which has 7 double values in total. But [Covariance](https://github.com/ros2/common_interfaces/blob/rolling/geometry_msgs/msg/PoseWithCovariance.msg) part adds 36 double values extra to the message.

And we generally don't need covariance of this size. We generally just have a variance in the 6 fields which make up the diagonal of the 6x6 matrix.

@isamu-takagi @yukkysaito @miursh what do you think? Should we create a new messages in `XWithVariance` format which have smaller footprint? And maybe even with variance part with 6 `float32`s.

Right now we are using [autoware_auto_perception_msgs/msg/DetectedObject.idl](https://github.com/tier4/autoware_auto_msgs/blob/tier4/main/autoware_auto_perception_msgs/msg/DetectedObject.idl) which add these pose, twist, accel with covariances which could be 7(xyz xyzw) + 6(xyzrpy errors) making 13 doubles per item. And for 3, it could be 13x3=39 doubles per Kinematics item. But right now it is (7+36)x3=129 doubles per item. 

And these messages with unnecessarily large memory footprints are passed around. And [when rviz2 reads them](https://github.com/autowarefoundation/autoware.universe/blob/main/common/autoware_auto_perception_rviz_plugin/src/object_detection/detected_objects_display.cpp#L49), I think it also causes lags with large object counts. I am also working on a benchmarking test node to prove it but it might take a while to work on it. https://www.youtube.com/watch?v=E86TWT03UXw here is a video from my initial attempts. I should probably create an issue to discuss lol.

Comment (mitsudome-r) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: I'm with Fatih on this.
I would like to know why we want these fields to be included for AD API. What are the use-cases of having covariance. (Same goes with `accelearation` and some of the fields in other message like `existence_probability` and `confidence`, although I don't care too much about those compared to covariance because they are relatively small in size)

Comment (yukkysaito) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: Thank you for comment.
Please wait a moment as I am preparing a simple figure to explain my opiniton.

Comment (yukkysaito) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: >And maybe even with variance part with 6 float32s.

If we do not allow rotation of the variance in the object coordinates, as in the image here, then six is fine.
image

If the following expressions are to be allowed, 7 or 8 variance value will be required.
image


Honestly, I too think 36 arrays are useless.
However, XXXwithCovariance is a standard format for ROS and has the advantage of being able to use some tools, so I am torn.

I also think the following is good.  I think never use position-rotation covariance.
image

Comment (mitsudome-r) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: We had some further discussion between @yukkysaito, @xmfcx and me.
At least we agree that covariance is unnecessary for AD API at all unless there is external software that needs covariance .

(Some off topic discussion)
We have also discussed about what we should do with [PredictedObjectKinetmatics](https://github.com/tier4/autoware_auto_msgs/blob/tier4/main/autoware_auto_perception_msgs/msg/PredictedObjectKinematics.idl) which is used inside Autoware. Our conclusion was that we would probably want to remove covariance for velocity and acceleration, and we could also remove unnecessary fields of covariance when we port the message to autoware_msgs. @xmfcx will create a discussion to discuss in further detail on this.

Comment (xmfcx) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: Alright, thanks a lot @yukkysaito 
I will create an issue and a sample message in the https://github.com/autowarefoundation/autoware_msgs to replace [autoware_auto_perception_msgs/msg/DetectedObject.idl](https://github.com/tier4/autoware_auto_msgs/blob/tier4/main/autoware_auto_perception_msgs/msg/DetectedObject.idl) with something simpler.

And for this PR, we can make it even simpler by removing covariances altogether. Because this is for visualization from the PR title.

And for the autoware_msgs, I have following covariance representations in my head:
Add a new message for covariance:
```
float32 var_pos_x
float32 var_pos_y
float32 var_pos_z
float32 var_rot_r
float32 var_rot_p
float32 var_rot_y
float32 cov_pos_xy
float32 cov_pos_xz
float32 cov_pos_yz
```

@yukkysaito what do you think?

> However, XXXwithCovariance is a standard format for ROS and has the advantage of being able to use some tools, so I am torn.

Yes, the ROS standard is to use `XWithCovariance` but do you have any libraries or visualization tools where this change would affect us negatively?

Most tools/libraries use Pose, Orientation parts from what I remember.

Comment (xmfcx) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: Also https://en.wikipedia.org/wiki/Covariance_matrix says:
> Any [covariance](https://en.wikipedia.org/wiki/Covariance) matrix is [symmetric](https://en.wikipedia.org/wiki/Symmetric_matrix) and [positive semi-definite](https://en.wikipedia.org/wiki/Positive_semi-definite_matrix) and its main diagonal contains [variances](https://en.wikipedia.org/wiki/Variance) (i.e., the covariance of each element with itself).

So half of the information is redundant/unnecessary, at worst, we only need half of the matrix anyway.

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/ObjectKinematics.msg:3]: @xmfcx @yukkysaito @mitsudome-r 
Thank you for your commend. Yes, I do agree for the adapi, we do not need to covariance as it will make the msg big.

> At least we agree that covariance is unnecessary for AD API at all unless there is external software that needs covariance .
I will confirm with one of the external software that going to use this api.

**Review by isamu-takagi — APPROVED** [review]
LGTM

File: autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg
Comment (xmfcx) [inline autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: I think it's better to add comment documentations for how `dimensions` and `footprint` are handled when different types are chosen.

**Review by xmfcx — CHANGES_REQUESTED** [review]

File: autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg
Comment (xmfcx) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: Something similar to https://github.com/ros2/common_interfaces/blob/humble/shape_msgs/msg/SolidPrimitive.msg

Comment (xmfcx) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: What about directly using the `SolidPrimitive.msg` message instead of added `ObjectShape.msg`?

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: @xmfcx Actually for this message, I cloned from the autoware perception message
It is the same message type
https://github.com/tier4/autoware_auto_msgs/blob/tier4/main/autoware_auto_perception_msgs/msg/Shape.idl


> What about directly using the SolidPrimitive.msg message instead of added ObjectShape.msg?

@yukkysaito @isamu-takagi How do you think?

Comment (xmfcx) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: Oh, we have contributed the missing features to the https://github.com/ros2/common_interfaces/blob/humble/shape_msgs/msg/SolidPrimitive.msg from Autoware.Auto. We were going to remove existing Shape.idl anyways. It's better to use the standard message.

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: I see, so autoware.universe planning to remove the shape.

I that's the case, I'm okay to use the `SolidPrimitive.msg`

Comment (yukkysaito) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: >I see, so autoware.universe planning to remove the shape.

This should be discussed a bit more.
Autoware's current Shape.idl is focused on the information needed to communicate in Perception-Planning. The interface should be limited to only the necessary information.
However, I think SolidPrimitive.msg is too expressive for general purpose.
Then, even when a Shape contains information that is not used in Autoware, it should be handled in a way that it works properly in Planning.

However, I am not opposed to using SolidPrimitive as an API.

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg:8]: Ya, sorry that should not be discuss at here.
For the api, we will use the `SolidPrimitive.msg`

Comment (tkhmy) [reply autoware_adapi_v1_msgs/perception/msg/ObjectShape.msg]: @yukkysaito @isamu-takagi @xmfcx 
Sorry for the long delay
I changed the shape to SolidPrimitive.msg at 7187d30
Please approve if it looks good for you

**Review by yukkysaito — APPROVED** [review]
LGTM

**Comment by tkhmy** [comment]
@xmfcx Hi, can you review the changes and approve if it looks fine for you? 
Thanks

**Review by xmfcx — APPROVED** [review]
Great work!

**Comment by tkhmy** [comment]
@isamu-takagi I do not have the merge permission, can you help me to merge it?

**Pull request merged**
